### ISSUE TITLE
toolChoice: 'required' and { toolName: 'x', type: 'tool' } send ai vercel SDK into endless loop when using streamText 

### ISSUE THREAD

**Issue by punkpeye** [vercel/ai#3944]
### Description

toolChoice: 'required' and { toolName: 'x', type: 'tool' } send ai vercel SDK into endless loop when using streamText 

### Code example

const { fullStream } = await streamText({
  abortSignal: signal,
  experimental_toolCallStreaming: false,
  maxSteps: 20,
  messages: relevantMessages.slice(),
  model
  toolChoice: { toolName: 'webSearch', type: 'tool' },
  tools: {
    webSearch: tool({
      description:
        'Search the web for information. Useful for when you need to answer questions about current events or a specific topic. The answer includes sources, which must be referenced in the answer using markdown footnotes.',
      execute: async ({ query }) => {
        return {
          answer: 'Paris',
        };
      },
      parameters: z.object({
        query: z
          .string()
          .describe(
            'The query to search for. Phrased in natural language as a question.',
          ),
      }),
    }),
  },
});

for await (const chunk of fullStream) {
      console.log('>>> chunk', chunk);
}

### AI provider

@ai-sdk/openai

### Additional context

Logs:

```
{
  type: 'tool-call',
  toolCallId: 'call_b0DkOmitAwd93Mt4xXd6u1gH',
  toolName: 'searchWeb',
  args: { query: 'What is the capital of France?' }
}
{
  type: 'tool-result',
  toolCallId: 'call_b0DkOmitAwd93Mt4xXd6u1gH',
  toolName: 'searchWeb',
  args: { query: 'What is the capital of France?' },
  result: {
    answer: 'The capital of France is Paris.',
    sources: [
      {
        name: 'What is the Capital of France? - WorldAtlas',
        snippet: "Learn about the history, geography, economy, tourism, and administration of Paris, the capital city of France and the country's largest city.",
        url: 'https://www.worldatlas.com/articles/what-is-the-capital-of-france.html'
      },
      {
        name: 'France | History, Maps, Flag, Population, Cities, Capital, & Facts ...',
        snippet: 'The capital and by far the most important city of France is Paris, one of the world’s preeminent cultural and commercial centres.',
        url: 'https://www.britannica.com/place/France'
      },
      {
        name: 'Paris - Wikipedia',
        snippet: 'Paris is the capital and largest city of France.',
        url: 'https://en.wikipedia.org/wiki/Paris'
      }
    ]
  }
}
{
  type: 'step-finish',
  finishReason: 'tool-calls',
  usage: { promptTokens: 456, completionTokens: 20, totalTokens: 476 },
  experimental_providerMetadata: { openai: { reasoningTokens: 0, cachedPromptTokens: 0 } },
  logprobs: undefined,
  response: {
    id: 'chatcmpl-AZpStkvY5DoWFRuCTA0L5hk1hOMOI',
    timestamp: 2024-12-02T01:24:35.000Z,
    modelId: 'gpt-4o-2024-11-20'
  },
  isContinued: false
}
{
  type: 'tool-call',
  toolCallId: 'call_4jKKKQAlZwguD1glEhCPAbDI',
  toolName: 'searchWeb',
  args: { query: 'capital of France' }
}
```

**Comment by lgrammel** [comment]
Of course, tool choice required means that you force tool calls. Together with maxSteps this enters such a loop. This is working as expected.

Can you describe what you want to achieve? there may be other approaches that could work here.

**Comment by punkpeye** [comment]
Suppose user sent a message: "What's the capital of France?"

User also ticked "Search", so I know their intent is to search.

<img width="291" alt="Screenshot 2024-12-02 at 8 29 17 AM" src="https://github.com/user-attachments/assets/bacaefb3-e43d-441b-a5c3-7654b4bd73bd">

I want to tell LLM that for this request it has to use `webSearch` tool.

Just leaving it as auto does not achieve the same result – it sometimes uses it, sometimes does not.

My understanding was that by setting tool to `required` or specific tool, I am telling LLM that to answer this question it must use a/the tool, but once the answer is resolved, it should not continue looping.

**Comment by lgrammel** [comment]
If you explicitly control that they want to use websearch through a flag, I recommend using a RAG approach vs a tool call approach.

Pseudo-code backend:
```ts
if (websearch) {
  const searchResult = await runMyWebsearch(...);
  prompt = buildMyPrompt(searchResult)
} else {
  prompt = myDefaultPrompt
}

return streamText( ... , with system or prompt influenced by the above, or by augmenting the last user msg)
```

**Comment by punkpeye** [comment]
> If you explicitly control that they want to use websearch through a flag, I recommend using a RAG approach vs a tool call approach.
> 
> Pseudo-code backend:
> 
> ```ts
> if (websearch) {
>   const searchResult = await runMyWebsearch(...);
>   prompt = buildMyPrompt(searchResult)
> } else {
>   prompt = myDefaultPrompt
> }
> 
> return streamText( ... , with system or prompt influenced by the above, or by augmenting the last user msg)
> ```

Do you happen to have any real-world examples of this?

**Comment by lgrammel** [comment]
The issue is that you specify the `toolChoice` for all steps, and it seems you only want to apply it to the first step. I'll think about a solution.

**Comment by brennanmceachran** [comment]
do we know if there is a way to turn it off after the first turn?

**Comment by arunegida** [comment]
if possible please have a feature that solves this problem of calling the tools no matter what, for now i have just tried a prompt, but i do think it will be good in long run

**Comment by demattosanthony** [comment]
> Tool choice required means that you force tool calls every time the LLM is called. Together with maxSteps this enters such a loop. This is working as expected.
> 
> Can you describe what you want to achieve? there may be other approaches that could work here.

i think this would be pretty nice to be able to control what tools are called at a certain step

**Comment by zachtball** [comment]
I solved this by doing:
```
const result = streamText({
      maxSteps: toolChoice === 'auto' ? maxSteps : 1,
...
```

But I agree, it would be really nice if this worked as expected. There are plenty of use cases for calling a tool once based on a user interaction.

> i think this would be pretty nice to be able to control what tools are called at a certain step

This is also a neat idea.

**Comment by lgrammel** [comment]
you can use prepareStep in ai sdk 5 for more flexibility, e.g. to force tool calls only in the first step

**Comment by BrentFarese** [comment]
Why would `toolChoice` required enter an infinite loop?  This seems like an anti-pattern.  Our use case is we have a triage/orchestration agent at a top level, that delegates requests to other sub-agents.  It would seem that `toolChoice: required` would force a tool(s) call **_until the request is satisfied_**.  For now, we are not using `toolChoice: required` and are having our system prompt mandate calling a tool(s), but it feels like we **_should_** be able to use `toolChoice: required` w/o an infinite loop being hit...

**Comment by gaieges** [comment]
I stumbled on this thread and found it to be helpful, but wanted to make it more explicit how I solved it.  @lgrammel has the right idea:



```typescript
    const result = streamText({
      model: openai("gpt-5-mini"),
      messages: convertToModelMessages(messages),
      system: systemPrompt,
      tools: {
        search: tool({
          description: "Search sites given a query.",
          inputSchema: z.object({
            query: z.string().describe("a single search query"),
          }),
          execute: doSearch,
        }),
      },
      // this will ensure that the first step of a new message will always require the tool call
      prepareStep: ({ stepNumber }) =>
        stepNumber === 1
          ? {
              toolChoice: { type: "tool", toolName: "search" },
            }
          : {},
    });
```

**Comment by sweetmantech** [comment]
This is my solution to force the toolCall for the first step with AI SDK v6 ToolLoopAgent class:
```
const agent = new ToolLoopAgent({
    model,
    instructions,
    tools,
    stopWhen,
    prepareStep: (options) => {
      return stepCountIs(1)
        ? {
            toolChoice: "required",
          }
        : options;
    },
  });
```

This forces an initial tool call for the first step, after which the LLM is free to continue calling tools or respond and stop the agent loop.

<img width="1440" height="900" alt="Image" src="https://github.com/user-attachments/assets/3f9d8e6c-1f4e-4dc5-8e17-30146f262648" />

**Comment by lgrammel** [comment]
## Identify issue type

Identified as **Feature**  
Confidence: **High**

### Rationale
This is a request for finer per-step control of tool selection rather than a failure of the documented `toolChoice: 'required'` behavior. The later `prepareStep` solution confirms it as an API flexibility enhancement.

### Evidence
- Maintainer states that `toolChoice` applies to every model call and the repeated calls with `maxSteps` are expected behavior.
- The requested behavior is to require a specific tool only on the first step, then allow the model to summarize or finish.
- The maintainer later identifies AI SDK 5's `prepareStep` as the mechanism for per-step tool choice.
